pkgs/development/interpreters/evcxr/default.nix

```diff
@@ -1,48 +1,57 @@
-{ cargo, fetchFromGitHub, makeWrapper, pkg-config, rustPlatform, lib, stdenv
-, gcc, cmake, libiconv, CoreServices, Security }:
-
-rustPlatform.buildRustPackage rec {
-  pname = "evcxr";
-  version = "0.15.0";
-
-  src = fetchFromGitHub {
-    owner = "google";
-    repo = "evcxr";
-    rev = "v${version}";
-    sha256 = "sha256-s8zM1vxEeJYcRek1rqUmrBfvB2zCAF3iLG8UVA7WABI=";
+{ callPackage
+, lib
+, makeWrapper
+, runCommand
+
+, CoreServices
+, Security
+
+, cargo
+, rustPlatform
+, rustc
+}:
+
+let
+  evcxr = callPackage ./evcxr.nix {
+    inherit CoreServices Security;
   };
 
-  cargoSha256 = "sha256-wMo5Fq6aMiE6kg8mZoz1T3KPwKSdJcej83MB+/GRM5w=";
-
-  RUST_SRC_PATH = "${rustPlatform.rustLibSrc}";
-
-  nativeBuildInputs = [ pkg-config makeWrapper cmake ];
-  buildInputs = lib.optionals stdenv.isDarwin
-    [ libiconv CoreServices Security ];
-
-  checkFlags = [
-    # test broken with rust 1.69:
-    # * https://github.com/evcxr/evcxr/issues/294
-    # * https://github.com/NixOS/nixpkgs/issues/229524
-    "--skip=check_for_errors"
-  ];
-
-  postInstall = let
-    wrap = exe: ''
-      wrapProgram $out/bin/${exe} \
-        --prefix PATH : ${lib.makeBinPath [ cargo gcc ]} \
-        --set-default RUST_SRC_PATH "$RUST_SRC_PATH"
-    '';
-  in ''
-    ${wrap "evcxr"}
-    ${wrap "evcxr_jupyter"}
-    rm $out/bin/testing_runtime
+  withPackages = callPackage ./withPackages.nix {
+    inherit cargo rustPlatform;
+  };
+
+  cargoHome = packageNames: runCommand "cargo-home" {} ''
+    mkdir -p $out/index_base
+    ln -s ${withPackages.cratesIndex} $out/index_base/index
+
+    mkdir -p $out
+
+    cp "${withPackages.cargoNix packageNames}"/Cargo.toml $out
+    cp "${withPackages.cargoNix packageNames}"/Cargo.lock $out
+
+    cat <<EOT >> $out/config.toml
+    [source.crates-io]
+    replace-with = "vendored-sources"
+
+    [source.vendored-sources]
+    directory = "${withPackages.vendorDependencies packageNames}"
+
+    [net]
+    offline = true
+    EOT
   '';
 
-  meta = with lib; {
-    description = "An evaluation context for Rust";
-    homepage = "https://github.com/google/evcxr";
-    license = licenses.asl20;
-    maintainers = with maintainers; [ protoben ma27 ];
-  };
-}
+in
+
+evcxr.overrideAttrs (oldAttrs: {
+  passthru = (oldAttrs.passthru or {}) // {
+    withPackages = packageNames:
+      runCommand "evcxr" { buildInputs = [makeWrapper]; } ''
+        mkdir -p $out/bin
+        makeWrapper ${evcxr}/bin/evcxr $out/bin/evcxr \
+          --set EVCXR_CONFIG_DIR ${withPackages.evcxrConfigDir packageNames} \
+          --set CARGO_HOME ${cargoHome packageNames} \
+          --prefix PATH : ${lib.makeBinPath [rustc cargo]}
+      '';
+    };
+})
```

Comment:
1. Can the `withPackages` function itself can be the sole export of `withPackages.nix`?
2. There is a pattern I have seen elsewhere that could be considered here. Consider [this snippet](https://github.com/NixOS/nixpkgs/blob/f74ac095c9ee174b221f72f50d5d47f2de427e40/pkgs/servers/monitoring/prometheus/sql-exporter.nix#L26), the passthru can reference the package itself passed in at the top of that file. This can work because the nix language is executed lazily. This can be leveraged to keep the main definition of evcxr that was currently extracted into evcxr.nix in the default.nix file.

Reply:
Okay, I've done this refactor and it seems to work. Certainly makes the PR nicer :)

I haven't used this pattern before, does it do the right thing with respect to overrides? For example, if I do `(evcxr.override {...}).withPackages [...]`, will the `withPackages` receive the overridden `evcxr`?

Reply:
That's an interesting question! I haven't tried it before, and my gut tells me no unless you override it using an overlay, but it could be worth trying.

Reply:
FWIW I recently learned about another pattern that might work here:

https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293/4